experimental/host/netlink/onewire.go

```diff
@@ -0,0 +1,552 @@
+		return nil, fmt.Errorf("invalid w1_netlink_msg status %d", status)
+	}
+	wantLen = getUint16(b[2:4])
+	b = b[12:]
+	if gotLen := len(b); gotLen != int(wantLen) {
+		return nil, fmt.Errorf("invalid w1_netlink_msg payload length %d, want %d", gotLen, wantLen)
+	}
+
+	return b, nil
+}
+
+// recvCmd performs a single read from the socket, verifies and removes netlink,
+// connector, W1 message, and W1 command headers, and returns the W1 command
+// payload. wantSeq and wantAck are the expected connector sequence and
+// acknowledgement numbers, respectively. wantMsgType and wantCmdType are the
+// expected W1 message and command types, respectively. An error is returned if
+// the received data does not match either of these values.
+func (s *socket) recvCmd(wantSeq, wantAck uint32, wantMsgType msgType, wantCmdType cmdType) ([]byte, error) {
+	b, err := s.recvMsg(wantSeq, wantAck, wantMsgType)
+	if err != nil {
+		return nil, err
+	}
+	if l := len(b); l < 4 {
+		return nil, fmt.Errorf("incomplete w1_netlink_cmd; got %d bytes, want at least 4", l)
+	}
+
+	// Check w1_netlink_cmd type and payload length.
+	if gotCmdType := cmdType(b[0]); gotCmdType != wantCmdType {
+		return nil, fmt.Errorf("invalid w1_netlink_cmd type %v, want %v", gotCmdType, wantCmdType)
+	}
+	wantLen := getUint16(b[2:4])
+	b = b[4:]
+	if gotLen := len(b); gotLen != int(wantLen) {
+		return nil, fmt.Errorf("invalid w1_netlink_cmd payload length %d, want %d", gotLen, wantLen)
+	}
+
+	return b, nil
+}
+
+func (s *socket) close() error {
+	fd := s.fd
+	s.fd = 0
+	return syscall.Close(fd)
+}
+
+//
+
+// driver1W implements periph.Driver.
+type driver1W struct {
+	buses []string
+}
+
+func (d *driver1W) String() string {
+	return "netlink-onewire"
+}
+
+func (d *driver1W) Prerequisites() []string {
+	return nil
+}
+
+func (d *driver1W) After() []string {
+	return nil
+}
+
+func (d *driver1W) Init() (bool, error) {
+	s, err := newSocket()
+	if err != nil {
+		return false, fmt.Errorf("netlink-onewire: failed to open socket: %v", err)
+	}
+	defer s.close()
+
+	// Find bus masters.
+	m := &w1Msg{typ: W1_LIST_MASTERS}
+	if err := s.sendMsg(m.serialize(), 0); err != nil {
+		return false, fmt.Errorf("netlink-onewire: failed to send list bus msg: %v", err)
+	}
+
+	b, err := s.recvMsg(0, 1, W1_LIST_MASTERS)
+	if err != nil {
+		return false, fmt.Errorf("netlink-onewire: failed to receive bus IDs: %v", err)
+	}
+
+	l := len(b)
+	if l%4 != 0 {
+		return false, fmt.Errorf("netlink-onewire: data size %d is not a multiple of 4", l)
+	}
+
+	var ids []uint32
+	for len(b) > 0 {
+		ids = append(ids, getUint32(b))
+		b = b[4:]
+	}
+
+	for _, id := range ids {
+		bus := int(id)
+		name := fmt.Sprintf("netlink-w1-master %d", bus)
+		d.buses = append(d.buses, name)
+		aliases := []string{fmt.Sprintf("OneWire%d", bus)}
+		if err := onewirereg.Register(name, aliases, bus, openerOneWire(bus).Open); err != nil {
+			return true, err
+		}
+	}
+	return true, nil
+}
+
+//
+
+func putUint16(b []byte, v uint16) {
```

Comment:
What about using binary.LittleEndian.PutUint16 ?
https://golang.org/pkg/encoding/binary/

Reply:
I was basically doing the same as in https://github.com/google/periph/blob/f4587c12845918ffaba8b966fd794753a4b84f5a/conn/onewire/onewire.go#L215. But binary.LittleEndian is of course fine.

Reply:
Ah, this particular case was because I worked on making sure that packages under conn/ do not depend on reflect, and "binary" does depend on reflect. But here you are already depending on fmt anyway.